### PR TITLE
feature: implemented text outline

### PR DESCRIPTION
**Pull request by Goldengide**
I noticed the stroke functionality on text watermark was not functional so I implemented that kindly merge please so i can see it in the next installment

### CODE REVIEW

File: src/Typography/Font.php [hunk 1]
```diff
@@ -11,6 +11,8 @@ class Font implements FontInterface
     protected float $size = 12;
     protected float $angle = 0;
     protected mixed $color = '000000';
+    protected mixed $strokeWidth = 1;
+    protected mixed $strokeColor = '000000';
```
Comment (olivervogel) [inline src/Typography/Font.php:15]: We already have black as font color. I suggest to set `ffffff` as default for stroke color.

File: src/Typography/Font.php [hunk 2]
```diff
@@ -11,6 +11,8 @@ class Font implements FontInterface
     protected float $size = 12;
     protected float $angle = 0;
     protected mixed $color = '000000';
+    protected mixed $strokeWidth = 1;
```
Comment (olivervogel) [inline src/Typography/Font.php:14]: By default the text should not be outline. So I guess we need to change this to `0`?

File: src/Interfaces/FontInterface.php [hunk 3]
```diff
@@ -13,13 +13,17 @@ interface FontInterface
      * @return FontInterface
      */
     public function setColor(mixed $color): self;
+    public function setStrokeWidth(mixed $width): self;
```
Comment (olivervogel) [inline src/Interfaces/FontInterface.php:16]: Please add proper doc blocks for new methods.

File: src/Interfaces/FontInterface.php [hunk 4]
```diff
@@ -13,13 +13,17 @@ interface FontInterface
...
+    public function setStrokeColor(mixed $color): self;
```
Comment (olivervogel) [inline src/Interfaces/FontInterface.php:17]: Please add proper doc blocks for new methods.

File: src/Interfaces/FontInterface.php [hunk 5]
```diff
@@ -13,13 +13,17 @@ interface FontInterface
...
 
     /**
      * Get color of font
      *
      * @return mixed
      */
     public function color(): mixed;
+    public function strokeWidth(): mixed;
```
Comment (olivervogel) [inline src/Interfaces/FontInterface.php:25]: Please add proper doc blocks for new methods.

File: src/Interfaces/FontInterface.php [hunk 6]
```diff
@@ -13,13 +13,17 @@ interface FontInterface
...
+    public function strokeColor(): mixed;
```
Comment (olivervogel) [inline src/Interfaces/FontInterface.php:26]: Please add proper doc blocks for new methods.

File: src/Typography/Font.php [hunk 7]
```diff
@@ -109,6 +111,19 @@ public function setColor(mixed $color): FontInterface
         return $this;
     }
 
+    public function setStrokeWidth(mixed $width): FontInterface
```
Comment (olivervogel) [inline src/Typography/Font.php:114]: Please add proper doc blocks for new methods.

File: src/Typography/Font.php [hunk 8]
```diff
@@ -109,6 +111,19 @@ public function setColor(mixed $color): FontInterface
...
+    {
+        $this->strokeWidth = $width;
+
+        return $this;
+    }
+    public function setStrokeColor(mixed $color): FontInterface
```
Comment (olivervogel) [inline src/Typography/Font.php:120]: Please add proper doc blocks for new methods.

File: src/Typography/Font.php [hunk 9]
```diff
@@ -119,6 +134,15 @@ public function color(): mixed
         return $this->color;
     }
 
+    public function strokeWidth(): mixed
```
Comment (olivervogel) [inline src/Typography/Font.php:137]: Please add proper doc blocks for new methods.

File: src/Typography/Font.php [hunk 10]
```diff
@@ -119,6 +134,15 @@ public function color(): mixed
...
+    {
+        return $this->strokeWidth;
+    }
+    public function strokeColor(): mixed
```
Comment (olivervogel) [inline src/Typography/Font.php:141]: Please add proper doc blocks for new methods.

File: src/Typography/FontFactory.php [hunk 11]
```diff
@@ -42,6 +42,19 @@ public function color(mixed $value): self
...
+
+        return $this;
+    }
+
+    public function strokeColor(mixed $color): self
+    {
+        $this->font->setStrokeColor($color);
+
+        return $this;
+    }
```
Comment (olivervogel) [inline src/Typography/FontFactory.php:57]: I suggest combining this into one method.

```php
public function stroke(int $width, mixed $color = 'ffffff'): self
```

**Review by olivervogel — CHANGES_REQUESTED** [review]
First of all, thanks for the pull request. I have tested the feature and it works fine. However, I have three important things that need to be added before I can integrate it.

1. The font system has been heavily rewritten by me in https://github.com/Intervention/image/pull/1292. I suggest that you build your PR on the state of [the not-yet-integrated state of #1292 ](https://github.com/Intervention/image/pull/1292), which will be integrated in the next minor release.

2. Intervention image is a library that provides a unified API for GD and Imagick. The goal is that every command will produce the same result when you replace the driver. I realize that this is not always 100% possible, as GD and Imagick work differently, but it should be a benchmark. However, the new feature currently only works with GD. I would ask you to add it for Imagick as well.

3. Some minor comments I wrote directly in the code.

File: src/Typography/Font.php (hunk 1 above)
Comment (Goldengide) [reply src/Typography/Font.php]: The font system has been heavily rewritten by me in https://github.com/Intervention/image/pull/1292. I suggest that you build your PR on the state of https://github.com/Intervention/image/pull/1292,

Please I sincerely don't know how to do this...

Comment (Goldengide) [reply src/Typography/Font.php]: Also I am not able to test imagegick so I couldn't update it to avoid bug and the whole code crashing

Comment (olivervogel) [reply src/Typography/Font.php]: I would suggest to merge [feature/text-wrapping](https://github.com/Intervention/image/tree/feature/text-wrapping) into this (your) feature branch and fix possible conflicts.

Comment (olivervogel) [reply src/Typography/Font.php]: Text outline with Imagick is definitely possible.

```php
$draw = new ImagickDraw();
$draw->setStrokeColor('#000');
$draw->setStrokeWidth(4);
$draw->setStrokeAntialias(true);
```

https://www.php.net/manual/en/imagickdraw.setstrokecolor.php
https://www.php.net/manual/en/imagickdraw.setstrokewidth.php
https://www.php.net/manual/en/imagickdraw.setstrokeantialias.php

And it's much easier than with GD. I would ask you to integrate it in your PR.

**Comment by olivervogel** [comment]
Thank you for your updates. Please edit the following two things because otherwise I cannot merge it.

1. The font system has been heavily rewritten by me in https://github.com/Intervention/image/pull/1292. I suggest that you build your PR on the state of [the not-yet-integrated state of #1292 ](https://github.com/Intervention/image/pull/1292), which will be integrated in the next minor release on February 17, 2024. I would suggest to merge [feature/text-wrapping](https://github.com/Intervention/image/tree/feature/text-wrapping) into this (your) feature branch and fix possible conflicts. After that you can push to GitHub again.

2. Intervention image is a library that provides a unified API for GD and Imagick. The goal is that every command will produce the same result when you replace the driver. I realize that this is not always 100% possible, as GD and Imagick work differently, but it should be a benchmark. However, the new feature currently only works with GD. I would ask you to add it for Imagick as well.

Text outline with Imagick is definitely possible.

```php
$draw = new ImagickDraw();
$draw->setStrokeColor('#000');
$draw->setStrokeWidth(4);
$draw->setStrokeAntialias(true);
```

https://www.php.net/manual/en/imagickdraw.setstrokecolor.php
https://www.php.net/manual/en/imagickdraw.setstrokewidth.php
https://www.php.net/manual/en/imagickdraw.setstrokeantialias.php

And it's even easier than with GD. 😄

**Comment by Goldengide** [comment]
merged feature/text-wrapping into add-outline-to-text branch

**Comment by Goldengide** [comment]
I am lost with the failed checks

**Review by olivervogel — CHANGES_REQUESTED** [review]

**Comment by Goldengide** [comment]
Please Sir I'd love to improve on my open source contribution what are the
things you feel I could improve and what other open source task can I take
on


Gideon Amowogbaje
[image: https://]about.me/amowogbaje
<https://about.me/amowogbaje?promo=email_sig&utm_source=product&utm_medium=email_sig&utm_campaign=edit_panel&utm_content=plaintext>

**Comment by olivervogel** [comment]
I just closed this because you opened another MR.

**Pull request closed without merging**